src/FastField.tsx

```diff
@@ -78,20 +75,77 @@ export class FastField<
 
   handleChange = (e: React.ChangeEvent<any>) => {
     e.persist();
-    const { validateOnChange } = this.context.formik;
+    const {
+      validateOnChange,
+      validate,
+      values,
+      validationSchema,
+    } = this.context.formik;
     const { type, value, checked } = e.target;
     const val = /number|range/.test(type)
       ? parseFloat(value)
       : /checkbox/.test(type) ? checked : value;
-    if (this.props.validate && validateOnChange) {
-      const maybePromise = (this.props.validate as any)(value);
-      if (isPromise(maybePromise)) {
-        this.setState({ value: val });
-        (maybePromise as any).then(noop, (error: string) =>
-          this.setState({ error })
-        );
+    if (validateOnChange) {
+      // Field-level validation
+      if (this.props.validate) {
+        const maybePromise = (this.props.validate as any)(value);
+        if (isPromise(maybePromise)) {
+          this.setState({ value: val });
+          (maybePromise as any).then(
+            () => this.setState({ error: undefined }),
+            (error: string) => this.setState({ error })
+          );
+        } else {
+          this.setState({ value: val, error: maybePromise });
+        }
+      } else if (validate) {
+        // Top-level validate
+        const maybePromise = (validate as any)(value);
+        if (isPromise(maybePromise)) {
+          this.setState({ value: val });
+          (maybePromise as any).then(
+            () => this.setState({ error: undefined }),
+            (error: any) =>
+              this.setState({ error: getIn(error, this.props.name) })
+          );
+        } else {
+          this.setState({
+            value: val,
+            error: getIn(maybePromise, this.props.name),
+          });
+        }
+      } else if (validationSchema) {
+        // Top-level validationsSchema
+        const schema = isFunction(validationSchema)
+          ? validationSchema()
+          : validationSchema;
+        const mergedValues = setIn(values, this.props.name, val);
+        // try to validate with yup synchronously if possible...saves a render.
+        try {
+          validateYupSchema(mergedValues, schema, true);
```

Comment:
This tries to validate synchronously with Yup. Yup distinguishes validation errors vs. other errors by error.name